README.md
# RaPhlot <img src = 'man/figures/logo.png' align = 'right' height = '110' />

For elegant political science graphs!


NEWS.md
## RaPhlot 0.1.0
* A brand new start!
---


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(RaPhlot)

test_check("RaPhlot")


tests/testthat/test-dydid.R
library(testthat)
library(RaPhlot)


R/dycoef.R
#' @title dycoef
#'
#' @description Running one regresison model over multiple subsamples by time periods, get the resuls, and plot coefficients over time.
#' @param data a data set object
#' @param dv  the outcome variable of your interest
#' @param x the explaining variable of your interest, should be available across several time periods
#' @param tid ID for each time point
#' @param years a sequence of perids you want to take into account; all years in your data by default.
#' @param covar control variables to be included
#'
#' @return
#' This functions return the following three objects:
#' A. "lastreg": regression results of you model in the subsample of last time period
#' B. "results": A data frame reporting coefficients and 95% CIs for all regressions.
#' C. "plot": A dynamic plot of x coefficients across periods using ggplot2().
#' @examples
#'
#' # load in your data
#' anes <- read.dta13("data/ANES cleaned.dta")
#'
#' # use the function
#' anes.dynamic <- dycoef(
#'   data = anes, dv = "pid7", x = 'ft_black', tid = "year",
#'   years = seq(1960,2020,4),
#'   covar = ('ft_white','as.factor(race)')
#' )
#'
#' # see the organized table for coefficients and 95% CIs
#' anes.dynamic$results
#'
#' # see the dynamic plot of DiD results
#' anes.dynamic$plot
#'
#' @export
#' @import tidyverse
#' @import lfe
#' @import fastDummies


## FUNCTION for coefficient trends
dycoef <- function(data,dv = 'pid7',x = 'ft_black',covar = NULL,tid,
                   years = seq(min(data[[tid]]),max(data[[tid]]),1),
                   linetype = 6, size = 3){

  # check var availability across time ---------------------------------

  ## function for checking
  tcheck <- function(data,var,tid){
    tcheck <- data %>%
      group_by(get(tid)) %>%
      summarize(dvm = mean(get(var), na.rm = TRUE)) %>%  # Use get(string) in dplyr to refer to the intended var
      filter(is.na(dvm) == FALSE)
    return(tcheck[['get(tid)']]) # Use data[[string]] to refer to the intended column?
  }

  ## check for x & y, then get the years that overlap
  # years <- seq(min(anes[[tid]]),max(anes[[tid]]),1)
  for (var in c(x,dv,covar) ){
    var <- ifelse(startsWith(var,"as.factor("),
                  sub("\\).*", "", sub( ".*as.factor\\(",'',var)),
                  var)

    yearsvar <- tcheck(data = data, var = var, tid = tid)

    years <- intersect(years, yearsvar)
  }


  # electionys <- seq(1952,2020,4) # whether to limit sample to election years
  # years <- intersect(years, electionys)
  print("Years where observations are available:")
  print(years)

  # set model and run regressions ---------------------------------------------------
  covar <- ifelse(is.null(covar), " ", paste("+", paste(covar, collapse = " + ")))
  model <- as.formula( paste(dv,'~',x,covar,'|0|0|0') )
  print(model)

  ## set a extraction function
  get.stats <- function(position, var, reg, time = NA) {
    stats <- data.frame(var = var,
                        coef = reg$coefficients[position],
                        se = reg$se[position], time = time)
    stats <- stats %>% mutate(ciup = coef + se * qnorm(.975), cilow = coef - se * qnorm(.975)) # get 95% CIs
    return(stats)
  }

  stats <- data.frame()

  for (i in years) {
    reg <- felm(model, data = data[data$year == i,])
    stats <- rbind(stats,get.stats(2,x,reg,i))
    # print(paste(i, '-- Regression Done'))
  }


  # plot and return --------------------------------------------------------------------

  dycoef.ribbon <- function(results) {
    plot <- ggplot(data = results, aes(x = time, y = coef, ymax = ciup, ymin = cilow)) + # geom_vline(xintercept = 2000,linetype = 2, size = 0.6, alpha = 0.5) +  # reference for any year
      geom_hline(yintercept = 0,linetype = 2, size = 0.6, alpha = 0.5) +  # reference for effect
      geom_ribbon(linetype = 2, fill = 'deepskyblue4' , alpha = 0.15) + # shaded area for 95% CIs
      geom_line(color = 'navyblue', linetype = linetype, alpha = 0.85) + # line and points for coefficients
      geom_point(color = 'navyblue', size = size) +
      theme_bw() +
      xlab('Year of Survey') +
      ylab('Point Estimate with 95% CIs') +
      scale_x_continuous(breaks = years) +
      theme(plot.title = element_text(face = 'bold',size = 12),
            legend.position = 'none',
            panel.grid.minor.x = element_blank(),
            axis.text = element_text(color = 'black'))

    return(plot)
  }

  # dycoef.lrange <- function(results) {
  #   plot <- ggplot(data = results, aes(x = time, y = coef, ymax = ciup, ymin = cilow)) + # geom_vline(xintercept = 2000,linetype = 2, size = 0.6, alpha = 0.5) +  # reference for any year
  #     geom_hline(yintercept = 0,linetype = 2, size = 0.6, alpha = 0.5) +  # reference for effect
  #     geom_linerange(color = 'slategray4', alpha = 0.50, size = 1.6) + # shaded area for 95% CIs
  #     geom_point(color = 'black', alpha = 0.9 ,size = 3)+ # aes(size = abs(coef)
  #     theme_bw() +
  #     xlab('Year of Survey') +
  #     ylab('Estimated Effect') + # scale_x_continuous(breaks = span[1]:span[2]) +
  #     theme(plot.title = element_text(face = 'bold',size = 12),
  #           legend.position = 'none')
  #   return(plot)
  # }

  return(list('plot' = dycoef.ribbon(stats), 'results' = stats,'lastreg' = summary(reg)))
}


R/dydid.R
#' @title dydid
#'
#' @description Set specifications for dynamic Difference-in-Differences analysis, report coefficients from regressions, and draw coefficients in a dynamic way.
#'
#' @param data a data set object
#' @param dv  the outcome variable of your interest
#' @param tpoint variable in your data that indicates the time when the treatment takes place
#' @param eid ID for each individual entity
#' @param tid ID for each time point
#' @param step the unit by which the time ID increments, e.g., 2 for every two years
#' @param span the number of periods before and after the treatment you want to take into consideration.
#' @param covar controls variables to be included
#'
#' @return
#' This functions return the following three objects:
#' A. "regs": Original results of DiD regressions using felm().
#' B. "results": A data frame reporting coefficients and 95% CIs for treatment-time indicators.
#' C. "plot": A dynamic plot of DiD coefficients using ggplot2().
#' @examples
#'
#' # load in your data
#' df <- read.dta13("data/big bad banks.dta")
#'
#' # use the function
#' bbb.did <- dydid(
#'   data = df, dv = "y", tpoint = "reform_timepoint", step = 1,
#'   eid = "state_id", tid = "year_id",
#'   span = c(-7, 7), covar = NULL
#' )
#'
#' # see the origigal regression results
#' bbb.did$regs
#'
#' # see the organized table for coefficients and 95% CIs
#' bbb.did$results
#'
#' # see the dynamic plot of DiD results
#' bbb.did$plot
#'
#' @export
#' @import tidyverse
#' @import lfe
#' @import fastDummies

# Prepare -----------------------------------------------------------------


# dydid: Function for Dynamic DiD Regression and Plot-------------------------------------------------------------


dydid <- function(data = df, dv = "y",
                  tpoint = "reform_timepoint", step = 1,
                  eid = "state_id", tid = "year_id", span = c(-5, 5), covar = NULL,
                  linetype = 5,size = 2.7) {

  # Set Treatment-Post -------------------------------------------
  df <- data

  df$relative <- df[, tid] - df[, tpoint]
  df <- df %>%
    mutate(post = relative / step) %>%
    mutate(pre = -relative / step)

  varnum <- span[2] - span[1]

  df$post <- ifelse(df$post >= span[2], span[2], df$post)
  df$post <- ifelse(df$post < 0, -1, df$post)
  df$pre <- ifelse(df$pre >= -span[1], -span[1], df$pre)
  df$pre <- ifelse(df$pre < 0, -1, df$pre)


  df <- fastDummies::dummy_cols(df, select_columns = "post")
  df <- fastDummies::dummy_cols(df, select_columns = "pre")

  pre_list <- paste0("pre_", -span[1]:2)
  post_list <- paste0("post_", 0:span[2])
  d_list <- c(pre_list, post_list)

  ## For observations without treatment, set all Dit indicators as 0
  for (i in d_list) {
    df[is.na(df[, tpoint]) == 1, i] <- 0
  }

  results <- rbind(
    data.frame(var = pre_list, time = span[1]:-2, position = 1:(0 - span[1] - 1)),
    data.frame(var = post_list, time = 0:span[2], position = -span[1]:varnum)
  )


  # DiD Regression ----------------------------------------------------------

  covar <- ifelse(is.null(covar), " ", paste("+", paste(covar, collapse = " + "))) # set formula part for covars

  model <- as.formula(
    paste(
      dv, "~", paste(pre_list, collapse = "+"), # remember to EXCLUDE pre_1
      "+", paste(post_list, collapse = "+"), # remember to INCLUDE post_0
      covar,
      "|", eid, "+", tid,
      "| 0 |", eid
    )
  ) # FE & Cluster SE

  rm(post_list, pre_list)

  did <- felm(model, data = df)


  # Extract Stats -----------------------------------------------------------
  ## set a extraction function

  get.stats <- function(position, var) {
    stats <- data.frame(var = var, coef = did$coefficients[position], se = did$cse[position])
    stats <- stats %>% mutate(ciup = coef + se * qnorm(.975), cilow = coef - se * qnorm(.975)) # get 95% CIs
    return(stats)
  }

  ## use loop to get stats for all indicators
  stats <- data.frame()
  for (i in 1:varnum) {
    stat <- get.stats(i, results[results$position == i, "var"])
    stats <- rbind(stats, stat)
  }

  ## combine variable lists and stats
  results <- left_join(results, stats, by = "var")
  rm(stats)

  # DiD Ploting Function-------------------------------------------------------------

  ## Recommended: CI as shade area
  did.plot <- function(results) {
    did_plot <- ggplot(data = results, aes(x = time, y = coef, ymax = ciup, ymin = cilow)) +
      geom_hline(yintercept = 0, linetype = 2, size = 0.6, alpha = 0.5) + # reference for effect
      geom_vline(xintercept = -1, linetype = 2, size = 0.6, alpha = 0.5) + # reference for post-treatment
      geom_ribbon(fill = "deepskyblue3", alpha = 0.15) + # shaded area for 95% CIs
      geom_line(color = "navyblue", linetype = linetype, alpha = 0.75) + # line and points for coefficients
      geom_point( color = "navyblue",size = size) +
      theme_bw() +
      xlab("Periods relative to Treatment") +
      ylab("Estimate of Treatment Effect") +
      scale_x_continuous(breaks = span[1]:span[2]) +
      theme(panel.grid.minor.x = element_blank(),
            legend.position = "none",
            axis.text = element_text(color = 'black')
      )
    return(did_plot)
  }

  # Plot and Return Results
  return(list("plot" = did.plot(results), "results" = results, "regs" = did))
}
